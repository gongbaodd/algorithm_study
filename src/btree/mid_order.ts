import { BTree } from "./lib";

function midorder(arr: number[], root: BTree<number> | null) {
  if (root === null) {
    return;
  }

  if (!root.lChild !== null) {
    midorder(arr, root.lChild);
  }

  arr.push(root.data as number);

  if (root.rChild !== null) {
    midorder(arr, root.rChild);
  }
}

export function midOrderToArr(root: BTree<number>) {
  const arr: number[] = [];

  midorder(arr, root);

  return arr;
}
