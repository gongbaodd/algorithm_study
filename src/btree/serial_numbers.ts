import { BTree } from "./lib";

export function toTree(
  arr: number[],
  start: number,
  end: number
): BTree<number> | null {
  let root: BTree<number> | null = new BTree<number>();

  if (end < start) {
    root = null;
    return root;
  }

  const mid = parseInt(((start + end) / 2).toString(), 10);

  root.data = arr[mid];

  root.lChild = toTree(arr, start, mid - 1);
  root.rChild = toTree(arr, mid + 1, end);

  return root;
}
