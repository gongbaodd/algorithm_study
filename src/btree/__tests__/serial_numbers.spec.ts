import { toTree } from "../serial_numbers";
import { midOrderToArr } from "../mid_order";
import { BTree } from "../lib";

describe("BTree", () => {
  describe("input [1,2,3,4,5,6,7,8,9,10]", () => {
    it("should return [1,2,3,4,5,6,7,8,9,10] in mid order", () => {
      const root = toTree([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0, 9);

      expect(midOrderToArr(root as BTree<number>)).toEqual([
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10
      ]);
      expect(root).toMatchSnapshot();
    });
  });
});
