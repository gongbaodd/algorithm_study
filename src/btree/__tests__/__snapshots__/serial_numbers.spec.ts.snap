// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BTree input [1,2,3,4,5,6,7,8,9,10] should return [1,2,3,4,5,6,7,8,9,10] in mid order 1`] = `
BTree {
  "data": 5,
  "lChild": BTree {
    "data": 2,
    "lChild": BTree {
      "data": 1,
      "lChild": null,
      "rChild": null,
    },
    "rChild": BTree {
      "data": 3,
      "lChild": null,
      "rChild": BTree {
        "data": 4,
        "lChild": null,
        "rChild": null,
      },
    },
  },
  "rChild": BTree {
    "data": 8,
    "lChild": BTree {
      "data": 6,
      "lChild": null,
      "rChild": BTree {
        "data": 7,
        "lChild": null,
        "rChild": null,
      },
    },
    "rChild": BTree {
      "data": 9,
      "lChild": null,
      "rChild": BTree {
        "data": 10,
        "lChild": null,
        "rChild": null,
      },
    },
  },
}
`;
